File: v4-client-py/examples/composite_example.py
"""Example for trading with human readable numbers

Usage: python -m examples.composite_example
"""
import asyncio
import logging
from random import randrange
from v4_client_py.chain.aerial.wallet import LocalWallet
from v4_client_py.clients import CompositeClient, Subaccount
from v4_client_py.clients.constants import BECH32_PREFIX, Network

from v4_client_py.clients.helpers.chain_helpers import (
    OrderType,
    OrderSide,
    OrderTimeInForce,
    OrderExecution,
)
from examples.utils import loadJson

from tests.constants import DYDX_TEST_MNEMONIC


async def main() -> None:
    wallet = LocalWallet.from_mnemonic(DYDX_TEST_MNEMONIC, BECH32_PREFIX)
    network = Network.config_network()
    client = CompositeClient(
        network,
    )
    subaccount = Subaccount(wallet, 0)
    ordersParams = loadJson("human_readable_orders.json")
    for orderParams in ordersParams:
        type = OrderType[orderParams["type"]]
        side = OrderSide[orderParams["side"]]
        time_in_force_string = orderParams.get("timeInForce", "GTT")
        time_in_force = OrderTimeInForce[time_in_force_string]
        price = orderParams.get("price", 1350)

        if time_in_force == OrderTimeInForce.GTT:
            time_in_force_seconds = 60
            good_til_block = 0
        else:
            latest_block = client.validator_client.get.latest_block()
            next_valid_block = latest_block.block.header.height + 1
            good_til_block = next_valid_block + 10
            time_in_force_seconds = 0

        post_only = orderParams.get("postOnly", False)
        try:
            tx = client.place_order(
                subaccount,
                market="ETH-USD",
                type=type,
                side=side,
                price=price,
                size=0.01,
                client_id=randrange(0, 100000000),
                time_in_force=time_in_force,
                good_til_block=good_til_block,
                good_til_time_in_seconds=time_in_force_seconds,
                execution=OrderExecution.DEFAULT,
                post_only=post_only,
                reduce_only=False,
            )
            print("**Order Tx**")
            print(tx)
        except Exception as error:
            print("**Order Failed**")
            print(str(error))

        await asyncio.sleep(5)  # wait for placeOrder to complete

    try:
        tx = client.place_order(
            subaccount,
            market="ETH-USD",
            type=OrderType.STOP_MARKET,
            side=OrderSide.SELL,
            price=900.0,
            size=0.01,
            client_id=randrange(0, 100000000),
            time_in_force=OrderTimeInForce.GTT,
            good_til_block=0,  # long term orders use GTBT
            good_til_time_in_seconds=1000,
            execution=OrderExecution.IOC,
            post_only=False,
            reduce_only=False,
            trigger_price=1000,
        )
        print("**Order Tx**")
        print(tx)
    except Exception as error:
        print("**Order Failed**")
        print(str(error))


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.get_event_loop().run_until_complete(main())

File: v4-client-js/examples/composite_example.ts
import { Order_TimeInForce } from '@dydxprotocol/v4-proto/src/codegen/dydxprotocol/clob/order';

import { BECH32_PREFIX } from '../src';
import { CompositeClient } from '../src/clients/composite-client';
import {
  Network, OrderSide,
} from '../src/clients/constants';
import LocalWallet from '../src/clients/modules/local-wallet';
import { Subaccount } from '../src/clients/subaccount';
import { randomInt } from '../src/lib/utils';
import { DYDX_TEST_MNEMONIC } from './constants';
import ordersParams from './human_readable_orders.json';

async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function test(): Promise<void> {
  const wallet = await LocalWallet.fromMnemonic(DYDX_TEST_MNEMONIC, BECH32_PREFIX);
  console.log(wallet);
  const network = Network.staging();
  const client = await CompositeClient.connect(network);
  console.log('**Client**');
  console.log(client);
  const subaccount = new Subaccount(wallet, 0);
  for (const orderParams of ordersParams) {
    try {
      const side = OrderSide[orderParams.side as keyof typeof OrderSide];
      const price = orderParams.price ?? 1350;

      const currentBlock = await client.validatorClient.get.latestBlockHeight();
      const nextValidBlockHeight = currentBlock + 1;
      // Note, you can change this to any number between `next_valid_block_height`
      // to `next_valid_block_height + SHORT_BLOCK_WINDOW`
      const goodTilBlock = nextValidBlockHeight + 3;

      const tx = await client.placeShortTermOrder(
        subaccount,
        'ETH-USD',
        side,
        price,
        0.01,
        randomInt(100_000_000),
        Order_TimeInForce.TIME_IN_FORCE_UNSPECIFIED,
        goodTilBlock,
        false,
      );
      console.log('**Order Tx**');
      console.log(tx);
    } catch (error) {
      console.log(error.message);
    }

    await sleep(5000);  // wait for placeOrder to complete
  }
}

test().then(() => {
}).catch((error) => {
  console.log(error.message);
});
